stub/src/main/java/io/grpc/stub/Calls.java

```diff
@@ -172,10 +172,11 @@
       ReqT param,
       Call.Listener<RespT> responseListener) {
     call.start(responseListener, new Metadata.Headers());
-    call.request(1);
     try {
       call.sendPayload(param);
       call.halfClose();
+      // Request after sending the payload otherwise we prematurely trigger a flush.
```

Comment:
A flush triggered by the call to 'request' can race with the flush in onReadComplete if the server responds quickly.

I have a change in mind that will make this moot anyway... Construct the deframer with pending messages == 1 rather than calling 'request' at the start of each call, this avoids a context switch for non-direct & blocking clients